core/src/main/java/hudson/util/ProcessTree.java

```diff
@@ -509,10 +532,51 @@ public void kill() throws InterruptedException {
             }
             
             LOGGER.log(FINER, "Killing {0}", getPid());
+            killSoftly(getPid());
             p.kill();
             killByKiller();
         }
 
+        private void killSoftly(int pid) {
+            // send Ctrl+C to the process in the first iteration
+            // after that just wait for it to cease to exist
+            long deadline = System.currentTimeMillis() + softkillWaitSeconds * 1000;
+            boolean sent = false;
+            do {
+                try {
+                    boolean exists = sendCtrlC(pid, sent);
+                    sent = true;
+                    if (!exists) {
+                        break;
+                    }
+                    TimeUnit.SECONDS.sleep(1);
+                }
+                catch (Exception e) {
+                    break;
+                }
+            } while(System.currentTimeMillis() < deadline);
```

Comment:
You want to use `System.nanoTime`, not `currentTimeMillis`. See Javadoc for why.

Reply:
I found that nanoTime is immune to system clock changes and hence preferred for elapsed-time calculations. Will fix!